### PR TITLE
FIX: propagation dependencies and rebuild out of date libs

### PR DESCRIPTION
**Pull request by gangliao**
Pain points:

When I implemented buddy allocator, finally I want to merge all small pieces libraries to `paddle_memory` for testing memory recycling effects. 

```bazel
if(${WITH_GPU})
  nv_library(system_allocator SRCS system_allocator.cc DEPS gflags cpu_info gpu_info)
else(${WITH_GPU})
  cc_library(system_allocator SRCS system_allocator.cc DEPS gflags cpu_info)
endif(${WITH_GPU})

cc_test(system_allocator_test SRCS system_allocator_test.cc DEPS system_allocator)

cc_library(meta_data SRCS meta_data.cc)

cc_library(meta_cache SRCS meta_cache.cc)

cc_library(memory_block SRCS memory_block.cc)

cc_library(buddy_allocator SRCS buddy_allocator.cc DEPS glog)

cc_library(memory SRCS memory.cc)

cc_library(paddle_memory
    DEPS
    memory meta_data
    meta_cache memory_block
    buddy_allocator system_allocator)

cc_test(memory_test SRCS memory_test.cc DEPS place paddle_memory)
```

I found the original generic implementation doesn't allow propagation dependencies and also cannot track out of date records.

The new features for merging static libraries.

1. propagation dependencies
2. rebuild the out of date static library

image

### CODE REVIEW

File: cmake/generic.cmake [hunk 1]
```diff
@@ -99,53 +99,72 @@ function(merge_static_libs TARGET_NAME)
...
+    target_link_libraries(${TARGET_NAME} ${libs_deps}) 
+
+    # Get the file name of the generated library
     set(outlibfile "$<TARGET_FILE:${TARGET_NAME}>")
 
     foreach(lib ${libs})
       add_custom_command(TARGET ${TARGET_NAME} POST_BUILD
-      COMMAND ${CMAKE_AR} ru ${outlibfile} @"../${lib}.objlist"
-      WORKING_DIRECTORY ${lib}.objdir)
+        COMMAND ${CMAKE_AR} cr ${outlibfile} *.o
```
Comment (typhoonzero) [inline cmake/generic.cmake]: Why not using modifier: `u`?

```
r[ab][f][u]  - replace existing or insert new file(s) into the archive
[u]          - only replace files that are newer than current archive contents
[c]          - do not warn if the library had to be created
```

Comment (gangliao) [inline cmake/generic.cmake]: https://bugzilla.redhat.com/show_bug.cgi?id=1155273

Comment (gangliao) [inline cmake/generic.cmake]: I tested this flag, it can add new symbols into the library.

Comment (typhoonzero) [inline cmake/generic.cmake]: Got it.

**Pull request merged**